File: SamplePresence/UIKit-ObjC/FormFieldCell.m
#import "FormFieldCell.h"

@implementation FormFieldCell

+ (NSString *)reuseIdentifier {
    return @"FormFieldCell";
}

- (instancetype)initWithStyle:(UITableViewCellStyle)style reuseIdentifier:(NSString *)reuseIdentifier {
    self = [super initWithStyle:style reuseIdentifier:reuseIdentifier];
    if (self) {
        self.selectionStyle = UITableViewCellSelectionStyleNone;
        _titleLabel = [[UILabel alloc] init];
        _titleLabel.font = [UIFont preferredFontForTextStyle:UIFontTextStyleBody];
        [_titleLabel setContentHuggingPriority:UILayoutPriorityRequired forAxis:UILayoutConstraintAxisHorizontal];
        [_titleLabel setContentCompressionResistancePriority:UILayoutPriorityRequired forAxis:UILayoutConstraintAxisHorizontal];

        _valueField = [[UITextField alloc] init];
        _valueField.borderStyle = UITextBorderStyleNone;
        _valueField.font = [UIFont preferredFontForTextStyle:UIFontTextStyleBody];
        _valueField.textAlignment = NSTextAlignmentRight;

        UIStackView *stack = [[UIStackView alloc] initWithArrangedSubviews:@[_titleLabel, _valueField]];
        stack.axis = UILayoutConstraintAxisHorizontal;
        stack.spacing = 12;
        stack.alignment = UIStackViewAlignmentCenter;
        stack.translatesAutoresizingMaskIntoConstraints = NO;
        [self.contentView addSubview:stack];

        UILayoutGuide *margins = self.contentView.layoutMarginsGuide;
        [NSLayoutConstraint activateConstraints:@[
            [stack.leadingAnchor constraintEqualToAnchor:margins.leadingAnchor],
            [stack.trailingAnchor constraintEqualToAnchor:margins.trailingAnchor],
            [stack.topAnchor constraintEqualToAnchor:self.contentView.topAnchor constant:10],
            [stack.bottomAnchor constraintEqualToAnchor:self.contentView.bottomAnchor constant:-10]
        ]];
    }
    return self;
}

- (void)configureWithTitle:(NSString *)title
                      text:(NSString *)text
               placeholder:(NSString *)placeholder
              keyboardType:(UIKeyboardType)keyboardType {
    self.titleLabel.text = title;
    self.valueField.text = text.length > 0 ? text : @"";
    self.valueField.placeholder = placeholder;
    self.valueField.keyboardType = keyboardType;
    self.valueField.autocapitalizationType = UITextAutocapitalizationTypeNone;
    self.valueField.autocorrectionType = UITextAutocorrectionTypeNo;
}

@end

@implementation FormMultilineCell

+ (NSString *)reuseIdentifier {
    return @"FormMultilineCell";
}

- (instancetype)initWithStyle:(UITableViewCellStyle)style reuseIdentifier:(NSString *)reuseIdentifier {
    self = [super initWithStyle:style reuseIdentifier:reuseIdentifier];
    if (self) {
        self.selectionStyle = UITableViewCellSelectionStyleNone;
        _titleLabel = [[UILabel alloc] init];
        _titleLabel.font = [UIFont preferredFontForTextStyle:UIFontTextStyleBody];

        _valueView = [[UITextView alloc] init];
        _valueView.font = [UIFont preferredFontForTextStyle:UIFontTextStyleBody];
        _valueView.scrollEnabled = NO;
        _valueView.delegate = self;
        _valueView.backgroundColor = UIColor.clearColor;

        UIStackView *stack = [[UIStackView alloc] initWithArrangedSubviews:@[_titleLabel, _valueView]];
        stack.axis = UILayoutConstraintAxisVertical;
        stack.spacing = 8;
        stack.translatesAutoresizingMaskIntoConstraints = NO;
        [self.contentView addSubview:stack];

        UILayoutGuide *margins = self.contentView.layoutMarginsGuide;
        [NSLayoutConstraint activateConstraints:@[
            [stack.leadingAnchor constraintEqualToAnchor:margins.leadingAnchor],
            [stack.trailingAnchor constraintEqualToAnchor:margins.trailingAnchor],
            [stack.topAnchor constraintEqualToAnchor:self.contentView.topAnchor constant:10],
            [stack.bottomAnchor constraintEqualToAnchor:self.contentView.bottomAnchor constant:-10],
            [_valueView.heightAnchor constraintGreaterThanOrEqualToConstant:80]
        ]];
    }
    return self;
}

- (void)configureWithTitle:(NSString *)title text:(NSString *)text {
    self.titleLabel.text = title;
    self.valueView.text = text;
}

- (void)textViewDidChange:(UITextView *)textView {
    if (self.onTextChanged) {
        self.onTextChanged(textView.text ?: @"");
    }
}

@end

@implementation FormSwitchCell

+ (NSString *)reuseIdentifier {
    return @"FormSwitchCell";
}

- (instancetype)initWithStyle:(UITableViewCellStyle)style reuseIdentifier:(NSString *)reuseIdentifier {
    self = [super initWithStyle:style reuseIdentifier:reuseIdentifier];
    if (self) {
        self.selectionStyle = UITableViewCellSelectionStyleNone;
        _titleLabel = [[UILabel alloc] init];
        _titleLabel.font = [UIFont preferredFontForTextStyle:UIFontTextStyleBody];
        _toggle = [[UISwitch alloc] init];

        UIStackView *stack = [[UIStackView alloc] initWithArrangedSubviews:@[_titleLabel, _toggle]];
        stack.axis = UILayoutConstraintAxisHorizontal;
        stack.spacing = 12;
        stack.alignment = UIStackViewAlignmentCenter;
        stack.translatesAutoresizingMaskIntoConstraints = NO;
        [self.contentView addSubview:stack];

        UILayoutGuide *margins = self.contentView.layoutMarginsGuide;
        [NSLayoutConstraint activateConstraints:@[
            [stack.leadingAnchor constraintEqualToAnchor:margins.leadingAnchor],
            [stack.trailingAnchor constraintEqualToAnchor:margins.trailingAnchor],
            [stack.topAnchor constraintEqualToAnchor:self.contentView.topAnchor constant:10],
            [stack.bottomAnchor constraintEqualToAnchor:self.contentView.bottomAnchor constant:-10]
        ]];
    }
    return self;
}

- (void)configureWithTitle:(NSString *)title isOn:(BOOL)isOn {
    self.titleLabel.text = title;
    self.toggle.on = isOn;
}

@end

File: SamplePresence/UIKit/FormFieldCell.swift
import UIKit

final class FormFieldCell: UITableViewCell {
    static let reuseIdentifier = "FormFieldCell"

    let titleLabel = UILabel()
    let valueField = UITextField()

    override init(style: UITableViewCell.CellStyle, reuseIdentifier: String?) {
        super.init(style: style, reuseIdentifier: reuseIdentifier)
        selectionStyle = .none

        titleLabel.font = .preferredFont(forTextStyle: .body)
        titleLabel.setContentHuggingPriority(.required, for: .horizontal)
        titleLabel.setContentCompressionResistancePriority(.required, for: .horizontal)

        valueField.borderStyle = .none
        valueField.font = .preferredFont(forTextStyle: .body)
        valueField.textAlignment = .right

        let stack = UIStackView(arrangedSubviews: [titleLabel, valueField])
        stack.axis = .horizontal
        stack.spacing = 12
        stack.alignment = .center
        stack.translatesAutoresizingMaskIntoConstraints = false
        contentView.addSubview(stack)

        NSLayoutConstraint.activate([
            stack.leadingAnchor.constraint(equalTo: contentView.layoutMarginsGuide.leadingAnchor),
            stack.trailingAnchor.constraint(equalTo: contentView.layoutMarginsGuide.trailingAnchor),
            stack.topAnchor.constraint(equalTo: contentView.topAnchor, constant: 10),
            stack.bottomAnchor.constraint(equalTo: contentView.bottomAnchor, constant: -10)
        ])
    }

    @available(*, unavailable)
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    func configure(title: String, text: String, placeholder: String, keyboardType: UIKeyboardType = .default) {
        titleLabel.text = title
        valueField.text = text
        valueField.placeholder = placeholder
        valueField.keyboardType = keyboardType
        valueField.autocapitalizationType = .none
        valueField.autocorrectionType = .no
    }
}

final class FormMultilineCell: UITableViewCell, UITextViewDelegate {
    static let reuseIdentifier = "FormMultilineCell"

    let titleLabel = UILabel()
    let valueView = UITextView()
    var onTextChanged: ((String) -> Void)?

    override init(style: UITableViewCell.CellStyle, reuseIdentifier: String?) {
        super.init(style: style, reuseIdentifier: reuseIdentifier)
        selectionStyle = .none

        titleLabel.font = .preferredFont(forTextStyle: .body)

        valueView.font = .preferredFont(forTextStyle: .body)
        valueView.isScrollEnabled = false
        valueView.delegate = self
        valueView.backgroundColor = .clear

        let stack = UIStackView(arrangedSubviews: [titleLabel, valueView])
        stack.axis = .vertical
        stack.spacing = 8
        stack.translatesAutoresizingMaskIntoConstraints = false
        contentView.addSubview(stack)

        NSLayoutConstraint.activate([
            stack.leadingAnchor.constraint(equalTo: contentView.layoutMarginsGuide.leadingAnchor),
            stack.trailingAnchor.constraint(equalTo: contentView.layoutMarginsGuide.trailingAnchor),
            stack.topAnchor.constraint(equalTo: contentView.topAnchor, constant: 10),
            stack.bottomAnchor.constraint(equalTo: contentView.bottomAnchor, constant: -10),
            valueView.heightAnchor.constraint(greaterThanOrEqualToConstant: 80)
        ])
    }

    @available(*, unavailable)
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    func configure(title: String, text: String) {
        titleLabel.text = title
        valueView.text = text
    }

    func textViewDidChange(_ textView: UITextView) {
        onTextChanged?(textView.text)
    }
}

final class FormSwitchCell: UITableViewCell {
    static let reuseIdentifier = "FormSwitchCell"

    let titleLabel = UILabel()
    let toggle = UISwitch()

    override init(style: UITableViewCell.CellStyle, reuseIdentifier: String?) {
        super.init(style: style, reuseIdentifier: reuseIdentifier)
        selectionStyle = .none

        titleLabel.font = .preferredFont(forTextStyle: .body)

        let stack = UIStackView(arrangedSubviews: [titleLabel, toggle])
        stack.axis = .horizontal
        stack.spacing = 12
        stack.alignment = .center
        stack.translatesAutoresizingMaskIntoConstraints = false
        contentView.addSubview(stack)

        NSLayoutConstraint.activate([
            stack.leadingAnchor.constraint(equalTo: contentView.layoutMarginsGuide.leadingAnchor),
            stack.trailingAnchor.constraint(equalTo: contentView.layoutMarginsGuide.trailingAnchor),
            stack.topAnchor.constraint(equalTo: contentView.topAnchor, constant: 10),
            stack.bottomAnchor.constraint(equalTo: contentView.bottomAnchor, constant: -10)
        ])
    }

    @available(*, unavailable)
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    func configure(title: String, isOn: Bool) {
        titleLabel.text = title
        toggle.isOn = isOn
    }
}
